File: test/language/array_types/python/FixedArrayUInt8Test.py
import os
import zserio

import ArrayTypes

from testutils import getApiDir

class FixedArrayUInt8Test(ArrayTypes.TestCase):
    def testBitSizeOf(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitPosition = 2
        fixedArrayBitSize = self.FIXED_ARRAY_LENGTH * 8
        self.assertEqual(fixedArrayBitSize, fixedArray.bitsizeof(bitPosition))

    def testInitializeOffsets(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + self.FIXED_ARRAY_LENGTH * 8
        self.assertEqual(expectedEndBitPosition, fixedArray.initialize_offsets(bitPosition))

    def testRead(self):
        writer = zserio.BitStreamWriter()
        self._writeFixedArrayToStream(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        fixedArray = self.api.FixedArray.from_reader(reader)

        uint8Array = fixedArray.uint8_array
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(uint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(i, uint8Array[i])

    def testWrite(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        bitBuffer = zserio.serialize(fixedArray)

        self.assertEqual(fixedArray.bitsizeof(), bitBuffer.bitsize)
        self.assertEqual(fixedArray.initialize_offsets(), bitBuffer.bitsize)

        readFixedArray = zserio.deserialize(self.api.FixedArray, bitBuffer)
        readUint8Array = readFixedArray.uint8_array
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(readUint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(i, readUint8Array[i])

    def testWriteReadFile(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH))
        fixedArray = self.api.FixedArray(uint8Array)
        zserio.serialize_to_file(fixedArray, self.BLOB_NAME)

        readFixedArray = zserio.deserialize_from_file(self.api.FixedArray, self.BLOB_NAME)
        readUint8Array = readFixedArray.uint8_array
        self.assertEqual(self.FIXED_ARRAY_LENGTH, len(readUint8Array))
        for i in range(self.FIXED_ARRAY_LENGTH):
            self.assertEqual(i, readUint8Array[i])

    def testWriteWrongArray(self):
        uint8Array = list(range(self.FIXED_ARRAY_LENGTH + 1))
        fixedArray = self.api.FixedArray(uint8_array_=uint8Array)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            fixedArray.write(writer)

    def _writeFixedArrayToStream(self, writer):
        for i in range(self.FIXED_ARRAY_LENGTH):
            writer.write_bits(i, 8)

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)), "fixed_array_uint8.blob")
    FIXED_ARRAY_LENGTH = 5

File: test/language/array_types/cpp/FixedArrayUInt8Test.cpp
#include "array_types/fixed_array_uint8/FixedArray.h"
#include "gtest/gtest.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace fixed_array_uint8
{

using allocator_type = FixedArray::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class FixedArrayUInt8Test : public ::testing::Test
{
protected:
    void writeFixedArrayToByteArray(zserio::BitStreamWriter& writer)
    {
        for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
            writer.writeBits(static_cast<uint32_t>(i), 8);
    }

    static const std::string BLOB_NAME;
    static const size_t FIXED_ARRAY_LENGTH = 5;
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::string FixedArrayUInt8Test::BLOB_NAME = "language/array_types/fixed_array_uint8.blob";

TEST_F(FixedArrayUInt8Test, bitSizeOf)
{
    vector_type<uint8_t> uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    const size_t bitPosition = 2;
    ASSERT_EQ(FIXED_ARRAY_LENGTH * 8, fixedArray.bitSizeOf(bitPosition));
}

TEST_F(FixedArrayUInt8Test, initializeOffsets)
{
    vector_type<uint8_t> uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    const size_t bitPosition = 2;
    ASSERT_EQ(bitPosition + FIXED_ARRAY_LENGTH * 8, fixedArray.initializeOffsets(bitPosition));
}

TEST_F(FixedArrayUInt8Test, readConstructor)
{
    zserio::BitStreamWriter writer(bitBuffer);
    writeFixedArrayToByteArray(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    FixedArray fixedArray(reader);

    const vector_type<uint8_t>& uint8Array = fixedArray.getUint8Array();
    const size_t numElements = FIXED_ARRAY_LENGTH;
    ASSERT_EQ(numElements, uint8Array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(i, uint8Array[i]);
}

TEST_F(FixedArrayUInt8Test, writeRead)
{
    vector_type<uint8_t> uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    zserio::BitStreamWriter writer(bitBuffer);
    fixedArray.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    FixedArray readFixedArray(reader);
    const vector_type<uint8_t>& readUint8Array = readFixedArray.getUint8Array();
    const size_t numElements = FIXED_ARRAY_LENGTH;
    ASSERT_EQ(numElements, readUint8Array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(i, readUint8Array[i]);
}

TEST_F(FixedArrayUInt8Test, writeReadFile)
{
    vector_type<uint8_t> uint8Array;
    uint8Array.reserve(FIXED_ARRAY_LENGTH);
    for (size_t i = 0; i < FIXED_ARRAY_LENGTH; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    zserio::serializeToFile(fixedArray, BLOB_NAME);

    FixedArray readFixedArray = zserio::deserializeFromFile<FixedArray>(BLOB_NAME);
    const vector_type<uint8_t>& readUint8Array = readFixedArray.getUint8Array();
    const size_t numElements = FIXED_ARRAY_LENGTH;
    ASSERT_EQ(numElements, readUint8Array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(i, readUint8Array[i]);
}

TEST_F(FixedArrayUInt8Test, writeWrongArray)
{
    vector_type<uint8_t> uint8Array;
    const size_t wrongArrayLength = FIXED_ARRAY_LENGTH + 1;
    uint8Array.reserve(wrongArrayLength);
    for (size_t i = 0; i < wrongArrayLength; ++i)
        uint8Array.push_back(static_cast<uint8_t>(i));
    FixedArray fixedArray;
    fixedArray.setUint8Array(uint8Array);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(fixedArray.write(writer), zserio::CppRuntimeException);
}

} // namespace fixed_array_uint8
} // namespace array_types
